sources/Context.cpp

```diff
@@ -163,162 +201,194 @@ Context::areInMainConnectedComponent(const std::vector<boost::shared_ptr<DYN::Bu
-  } else if (type == "BRANCH") {
-    bool r = checkLine(id) || checkTwoWTransformer(id);
-    if (!r) {
-      return std::string("not found as line or two-windings transformer");
+std::vector<Contingencies::ElementInvalidReason>
+Context::checkContingency(std::shared_ptr<dfl::inputs::Contingencies::ContingencyDefinition> c) const {
+  std::vector<Contingencies::ElementInvalidReason> result;
+  LOG(debug) << c->id << LOG_ENDL;
+
+  // Check each element and see whether they are good for inclusion
+  for (auto e: c->elements) {
+    LOG(debug) << "  " << e.id << " (" << Contingencies::toString(e.type) << ")" << LOG_ENDL;
+    auto invalid = checkContingencyElement(e.id, e.type);
+    if (invalid) {
+      LOG(warn) << MESS(ElementInvalid, e.id, Contingencies::toString(e.type), Contingencies::toString(*invalid)) << LOG_ENDL;
+      result.push_back(*invalid);
+    } else {
+      LOG(debug) << "  Element " << e.id << "(" << Contingencies::toString(e.type) << ") is valid" << LOG_ENDL;
     }
-    return boost::none; // No problem
-  } else if (type == "SHUNT_COMPENSATOR") {
-    return checkShuntCompensator(id);
-  } else if (type == "LOAD") {
-    return checkLoad(id);
-  } else if (type == "DANGLING_LINE") {
-    return checkDanglingLine(id);
-  } else if (type == "STATIC_VAR_COMPENSATOR") {
-    return checkStaticVarCompensator(id);
-  } else if (type == "BUSBAR_SECTION") {
-    return boost::none;
   }
 
-  // FIXME complete access to the different types of elements through NetworkInterface
-  return std::string("type not known"); // No problem
+  return result;
 }
 
-std::vector<std::string>
-Context::checkContingencies() const {
-  std::vector<std::string> result;
+std::tuple<
+  std::vector<std::shared_ptr<dfl::inputs::Contingencies::ContingencyDefinition>>,
+  std::vector<dfl::inputs::Contingencies::ElementInvalidReason>
+>
+Context::checkAndFilterContingencies() const {
+  std::vector<Contingencies::ElementInvalidReason> errors;
   LOG(debug) << "Contingencies. Check that all elements of contingencies are valid for disconnection simulation" << LOG_ENDL;
-  const auto& contingencies = contingencies_.definitions();
-  for (auto c = contingencies.begin(); c != contingencies.end(); ++c) {
-    LOG(debug) << c->id << LOG_ENDL;
-    for (auto e = c->elements.begin(); e != c->elements.end(); ++e) {
-      LOG(debug) << "  " << e->id << " (" << e->type << ")" << LOG_ENDL;
-      std::string reason;
-      auto invalid = checkContingencyElement(e->id, e->type);
-      if (invalid) {
-        LOG(warn) << "  Element " << e->id << " (" << e->type << ") not valid, reason: " << invalid.value() << LOG_ENDL;
-        result.push_back(invalid.value());
-      } else {
-        LOG(debug) << "  Element " << e->id << "(" << e->type << ") is valid" << LOG_ENDL;
-      }
+
+  // We'll store here the iterators we don't want
+  std::vector<std::vector<std::shared_ptr<dfl::inputs::Contingencies::ContingencyDefinition>>::iterator> iteratorsToRemove;
```

Comment:
Agree, though it is a bit complicated in C++. Functional programming is very verbose and kinda clunky, I event gave it a try, but it was more difficult to read, there's also the problem that we are doing a filter and a fold at the same time (since we are getting the errors).
The good news is that the functional spirit is still preserved, we actually operate on a copy and the whole method is `const`.